### PR TITLE
New specs/version for Erlang and CouchDB 3.1.2

### PR DESCRIPTION
**Pull request by amaltaro**
Most of this work has been done by Max Uetrecht (thanks!), including the porting of all our couchdb patches to CouchDB 3.1. A BIG thank you to Shahzad as well for the priceless supporting to build these specs.

A short summary of changes is:
* remove WMCore services dependency on `couchdb16`, instead define it to depend on `couchdb31`
* stop building `couchdb15` in the comp meta-package
* created new spec for CouchDB 3.1.2 version
* port replicator, ssl and cms-auth patches to CouchDB 3.1
* created new spec for Erlang 22.3.4.19
* port Erlang patches
* update WMAgent to `2.0.4.patch1` and make it depend on the new Couch version

More info at: https://git.phenomax.me/phenomax/cern-couchdbv3-rpm/-/blob/master/couchdb311.spec

Other references for Erlang and CouchDB 3.1 can be found at:
https://github.com/rabbitmq/erlang-rpm/blob/erlang-22.3.x/erlang.spec
https://github.com/adrienverge/copr-couchdb/blob/master/couchdb.spec

### CODE REVIEW

**Comment by muhammadimranfarooqi** [comment]
+1 to 3e53af65a61f5b39a6f8a9d9b5534f8ed5d62dcc on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/7218/test_3e53af6-20210818-070502.log" target="_blank">here</a>.

File: couchdb31.spec [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+%patch0 -p0
+%patch1 -p0
+#mv %_sourcedir/couchdb31_cms_auth.erl %_sourcedir/couch_cms_auth.erl
+#ls -l %_builddir/apache-couchdb-%realversion/src/couch
+cp %_sourcedir/couchdb31_cms_auth.erl %_builddir/apache-couchdb-%{realversion}/src/couch/src/couch_cms_auth.erl
+
+%build
+# apache 1.5.1 does not have option to specify --with-icu4c, instead
+# they used --with-win32-icu-binaries which mostly the same
+export CURL_ROOT SPIDERMONKEY_ROOT ICU4C_ROOT ERLANG_ROOT AUTOTOOLS_ROOT
```
Comment (smuzaffar) [inline couchdb31.spec]: @amaltaro , please use ERLANG22_ROOT every where in this spec as you depend on `erlang22`

Comment (amaltaro) [reply couchdb31.spec]: Thanks for pointing this out, Shahzad. I have it fixed in the second commit.

**Comment by amaltaro** [comment]
@smuzaffar Shahzad, with your help, I managed to have some progress on this build, but now the couchdb build fails to find this C header:
```
Compiling /build/amaltaro/w/BUILD/slc7_amd64_gcc630/external/couchdb31/3.1.1/apache-couchdb-3.1.1/src/couch/priv/couch_js/1.8.5/http.c
/build/amaltaro/w/BUILD/slc7_amd64_gcc630/external/couchdb31/3.1.1/apache-couchdb-3.1.1/src/couch/priv/couch_js/1.8.5/http.c:18:19: fatal error: jsapi.h: No such file or directory
 #include <jsapi.h>
                   ^
compilation terminated.
```

full logs in vocms055: /build/amaltaro/w/BUILD/slc7_amd64_gcc630/external/couchdb31/3.1.1/log

As you can see in the new Couch spec, I tried setting the JS_CFLAGS/JS_LDFLAGS, both explicitly defining ".../include" or ".../include/js", but still it does not work.

I can confirm that that library is shipped with spidermonkey:
```
cmst1@vocms0192:/data/srv/wmagent/current $ ls sw/slc7_amd64_gcc630/external/spidermonkey/1.8.5-comp4/include/js/jsapi.h 
sw/slc7_amd64_gcc630/external/spidermonkey/1.8.5-comp4/include/js/jsapi.h
```

yet I fail to find it during build time.

For the record, I'm trying these steps based on this comment:
https://github.com/apache/couchdb/issues/455#issuecomment-287523718

Would you have any suggestion on how I could proceed here? If you prefer, we can stick to that private email thread as well ;)

**Comment by muhammadimranfarooqi** [comment]
+1 to 6736183fa755c6b326c84d4067faaf95d1dcdbf9 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/7218/test_6736183-20210818-103503.log" target="_blank">here</a>.

**Comment by amaltaro** [comment]
Nevermind, Shahzad. I can "patch" that file with something like:
```
sed -i "s+-I/usr/include/js -I/usr/local/include/js+-I$SPIDERMONKEY_ROOT/include/js+" %_builddir/apache-couchdb-%{realversion}/src/couch/rebar.config.script
```
and the build moves on. Thanks.

**Comment by smuzaffar** [comment]
yes you can do that but there should be a way to pass this information. Did you try setting 
JS_CFLAGS and JS_LDFLAGS as environment variables? Try setting these before calling `configure`

**Comment by amaltaro** [comment]
Shahzad, I will try setting these flags once I manage to build it with the current changes.
And I believe I got to the very end, but now it fails with a message like:
"""
    You can now copy the rel/couchdb directory anywhere on your system.
    Start CouchDB with ./bin/couchdb from within that directory.


Notice: There is no 'make install' command for CouchDB 2.x+.

    To install CouchDB into your system, copy the rel/couchdb
    to your desired installation location. For example:
    cp -r rel/couchdb /usr/local/lib
"""

so I need to remove the "make install" and just copy data around. Can you please clarify where I should take that `rel/couchdb` from? Is it from `%_builddir`? And I do not know where I should copy it to as well, is it going to be `%{i}`?
@smuzaffar

**Comment by smuzaffar** [comment]
it should be under %{_builddir}/apache-couchdb-%{realversion}

**Comment by smuzaffar** [comment]
and yes you should copy it under %{i}. you might want it to be under %{i}/bin

**Comment by muhammadimranfarooqi** [comment]
+1 to a08398638328622690dcf3ac58b8381eeee5da5a on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/7218/test_a083986-20210818-120503.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
+1 to d9c3f03b9af0c81e5ff00c269bcd8050d1379457 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/7218/test_d9c3f03-20210818-131503.log" target="_blank">here</a>.

**Comment by amaltaro** [comment]
Okay, I tried replacing these lines:
```
sed -i "s+-I/usr/include/js -I/usr/local/include/js+-I$SPIDERMONKEY_ROOT/include/js -I$ICU4C_ROOT/include+" %_builddir/apache-couchdb-%{realversion}/src/couch/rebar.config.script
sed -i "s+-L/usr/local/lib -lmozjs185+-L$SPIDERMONKEY_ROOT/lib -lmozjs185+" %_builddir/apache-couchdb-%{realversion}/src/couch/rebar.config.script
```

by
```
export JS_CFLAGS="-I$SPIDERMONKEY_ROOT/include/js -I$ICU4C_ROOT/include"
export JS_LDFLAGS="-L$SPIDERMONKEY_ROOT/lib -lmozjs18 -lm -L$ICU4C_ROOT/lib"
```

before running ./configure... , but I had no success and it fails to compile the same file previously mentioned (...include/js/jsapi.h).

I am now able to build couchdb with the current spec. I'm going to squash these commits and cleanup the debugging I had.
If you have any other improvements/recommendations, please let me know Shahzad.

**Comment by muhammadimranfarooqi** [comment]
+1 to 26a205d16a7c1671679a8fcaf7397e54decb0335 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/7218/test_26a205d-20210818-142003.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
+1 to ad284c34d3848f7b8ad6687733584db08fc17c5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/7218/test_ad284c3-20210820-134502.log" target="_blank">here</a>.

**Comment by amaltaro** [comment]
It turns out the `couchdb` binary wasn't getting added to the PATH, because the replacement for `make install` (cp command) wasn't copying the couchdb code with the proper directory structure.

Second commit fixes that issue, and a new docker image has been created and is apparently working fine now.

**Comment by muhammadimranfarooqi** [comment]
to 599d8e3de8abc33eaefe3d47e72044b17a5682f3 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/7218/test_599d8e3-20211111-121504.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
to 599d8e3de8abc33eaefe3d47e72044b17a5682f3 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/7218/test_599d8e3-20220214-112503.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
to 599d8e3de8abc33eaefe3d47e72044b17a5682f3 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/7218/test_599d8e3-20220214-145004.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
to 599d8e3de8abc33eaefe3d47e72044b17a5682f3 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/7218/test_599d8e3-20220214-172503.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
to 599d8e3de8abc33eaefe3d47e72044b17a5682f3 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/7218/test_599d8e3-20220215-101004.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
+1 to 599d8e3de8abc33eaefe3d47e72044b17a5682f3 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/7218/test_599d8e3-20220215-105504.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
to 9ebf9b7bbb5c03b9fd31b3da55a0d232a2532a30 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/7218/test_9ebf9b7-20220225-025503.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
+1 to 9ebf9b7bbb5c03b9fd31b3da55a0d232a2532a30 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/7218/test_9ebf9b7-20220225-084003.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to e6e35df442f4dc6b125e05968bdf7efa4a3bdc7e on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/7218/test_e6e35df-20220311-211004.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to a667384e6971899410b8583e476d93338cf6ee15 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/7218/test_a667384-20220311-214004.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
+1 to a4b1ae4c1a35707d01bda657d855df295f3c2e64 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/7218/test_a4b1ae4-20220312-010503.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
+1 to 3c1a388b020008f6d726d40c23b4432a91605103 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/7218/test_3c1a388-20220610-045003.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
+1 to b31b21fb5002c9a061eeeced7e2d5d26cb4fb751 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/7218/test_b31b21f-20220616-062003.log" target="_blank">here</a>.

**Comment by amaltaro** [comment]
These changes have been largely tested now. Before we can deploy central services, further tweaks are required for the frontend code though.

**Comment by amaltaro** [comment]
merge

**Pull request merged**